src/wrapper/java_vm/vm.rs

```diff
@@ -132,9 +190,128 @@ impl<'a> Deref for AttachGuard<'a> {
 }
 
 impl<'a> Drop for AttachGuard<'a> {
+    fn drop(&mut self) {
+        if self.should_detach {
+            InternalAttachGuard::clear_tls();
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum ThreadType {
+    Normal,
+    Daemon,
+}
+
+#[derive(Debug)]
+struct InternalAttachGuard {
+    java_vm: *mut sys::JavaVM,
+}
+
+impl InternalAttachGuard {
+    fn create_and_attach(
+        java_vm: *mut sys::JavaVM,
+        ty: ThreadType,
+    ) -> Result<*mut sys::JNIEnv> {
+        let guard = InternalAttachGuard::new(java_vm);
+        let env_ptr = unsafe {
+            if ty == ThreadType::Daemon {
+                guard.attach_current_thread_as_daemon()?
+            } else {
+                guard.attach_current_thread()?
+            }
+        };
+
+        Self::fill_tls(guard);
+
+        Ok(env_ptr)
+    }
+
+    fn new(java_vm: *mut sys::JavaVM) -> Self {
+        Self {
+            java_vm,
+        }
+    }
+
+    /// Stores guard in thread local storage.
+    fn fill_tls(guard: InternalAttachGuard) {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = Some(guard);
+        });
+    }
+
+    /// Clears thread local storage, drops InternalAttachGuard and causes detach.
+    fn clear_tls() {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = None;
+        });
+    }
+
+    unsafe fn attach_current_thread(&self) -> Result<*mut sys::JNIEnv> {
+        let mut env_ptr = ptr::null_mut();
+        let res = java_vm_unchecked!(
+            self.java_vm,
+            AttachCurrentThread,
+            &mut env_ptr,
+            ptr::null_mut()
+        );
+        jni_error_code_to_result(res)?;
+
+        ATTACHED_THREADS.fetch_add(1, Ordering::SeqCst);
+
+        debug!("Attached thread {:?}. Name: {:?}. {} threads attached",
+               current().id(),
+               current().name(),
+               ATTACHED_THREADS.load(Ordering::SeqCst)
+        );
+
+        Ok(env_ptr as *mut sys::JNIEnv)
+    }
+
+    unsafe fn attach_current_thread_as_daemon(&self) -> Result<*mut sys::JNIEnv> {
+        let mut env_ptr = ptr::null_mut();
+        let res = java_vm_unchecked!(
+            self.java_vm,
+            AttachCurrentThreadAsDaemon,
+            &mut env_ptr,
+            ptr::null_mut()
+        );
+        jni_error_code_to_result(res)?;
+
+        ATTACHED_THREADS.fetch_add(1, Ordering::SeqCst);
+
+        debug!("Attached daemon thread {:?}. Name: {:?}. {} threads attached",
+               current().id(),
+               current().name(),
+               ATTACHED_THREADS.load(Ordering::SeqCst)
+        );
+
+        Ok(env_ptr as *mut sys::JNIEnv)
+    }
+
+    fn detach(&mut self) -> Result<()> {
+        unsafe {
+            java_vm_unchecked!(self.java_vm, DetachCurrentThread);
+        }
+        ATTACHED_THREADS.fetch_sub(1, Ordering::SeqCst);
```

Comment:
Is that correct that it won't go below zero even with a public  `detach_current_thread` because **all** detaches happen through this method?

Reply:
Yes, all detaches somehow triggers IAG::drop, which is the only entry points for this method